sled-agent/src/opte.rs

```diff
@@ -0,0 +1,334 @@
+        // VPC subnet.
+        match subnet.network() {
+            IpAddr::V4(ip) => {
+                let prefix =
+                    Ipv4PrefixLen::new(subnet.prefix()).map_err(|e| {
+                        opte_ioctl::Error::InvalidArgument(format!(
+                            "Invalid IPv4 subnet prefix: {}",
+                            e
+                        ))
+                    })?;
+                let cidr = Ipv4Cidr::new(opte::api::Ipv4Addr::from(ip), prefix);
+                let route = AddRouterEntryIpv4Req {
+                    port_name: name.clone(),
+                    dest: cidr,
+                    target: RouterTarget::VpcSubnet(IpCidr::Ip4(cidr)),
+                };
+                hdl.add_router_entry_ip4(&route)?;
+            }
+            IpAddr::V6(_) => {
+                return Err(opte_ioctl::Error::InvalidArgument(String::from(
+                    "IPv6 not yet supported",
+                ))
+                .into());
+            }
+        }
+
+        // Create a VNIC on top of this device, to hook Viona into.
+        //
+        // Viona is the virtio-net driver used to provide virtual interfaces to
+        // guests. It sits on top of a MAC provider, which is responsible for
+        // delivering frames to the underlying data link. In theory, Viona can
+        // work with any MAC provider. However, there are several implicit
+        // assumptions, in both Viona _and_ MAC, that require Viona to be built
+        // on top of a VNIC specifically. There is probably a good deal of work
+        // required to relax that assumption, so in the meantime, we create a
+        // superfluous VNIC on the OPTE device, solely so Viona can use it.
+        let vnic = {
+            let phys = PhysicalLink(name.clone());
+            let vnic_name = format!("v{}", name);
+            Dladm::create_vnic(
+                &phys,
+                &vnic_name,
+                Some(omicron_common::api::external::MacAddr(mac)),
+                None,
+            )?;
+            Some(Vnic::wrap_existing(vnic_name))
+        };
+
+        Ok(OptePort {
+            name,
+            ip,
+            subnet,
+            mac,
+            vni,
+            underlay_ip,
+            gateway,
+            boundary_services,
+            vnic,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct BoundaryServices {
+    pub ip: Ipv6Addr,
+    pub vni: Vni,
+}
+
+impl Default for BoundaryServices {
+    fn default() -> Self {
+        // TODO-completeness: Don't hardcode this.
+        //
+        // Boundary Services will be started on several sleds during rack
+        // setup, and the addresses of the chosen sleds will need to be
+        // propagated here.
+        const BOUNDARY_SERVICES_ADDR: Ipv6Addr =
+            Ipv6Addr::new(0xfd00, 0x99, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01);
+        let boundary_services_vni = Vni::new(99_u32).unwrap();
```

Comment:
My initial thought here would be to have some relatively small reserved space, something like 0-1024 for things like boundary services, and have the rest of the 2^24 sized space be allocated with a simple linear counter with rollover. At 1000 VNI allocations per day it would take approximately 45 years to roll over. We could also play it safe from a rollover perspective and keep a list that backfills VNIs as they get recycled, and refuses to over-allocate. I could see a situation arising where a bug or a client program starts allocating in a tight loop and we find ourselves in a colliding rollover situation.

I do think we want to be cautious about loading too much semantic meaning atop a VNI and just let the VNI be a VNI. We can always have other indices that provide greater context around a VNI.

Reply:
I must have misunderstood something deeply. I was under the impression the VNI is effectively an identifier for the VPC. Is that not accurate? If it _is_ accurate, we need a deterministic way to derive the VNI from the VPC, which is why I'm currently taking the first 24 bits of the VPC UUID.

Reply:
I believe the VNI is an identifier for the VPC. I'm not sure how safe truncating a unique (128 bit?) uuid/hash to 24 bits is. I could see a situation where we have unique VPC UUIDs but non-unique VNIs derived from some 24 bit part of that UUID.

Reply:
Ah, sorry if I was unclear. The current solution of truncating the UUID is _not_ correct. It was a simple, temporary solution. I was asking about your suggestion that we sequentially allocate VNIs. I had interpreted that as being done by the sled agent, which I now see is probably not what you meant. I think we either need (1) a sequential or unique VNI allocated to a VPC by the database at creation time or (2) a deterministic way to derive the VNI from the VPC on each sled when the OPTE port is created.

I think (1) might be closer to what you had in mind. I'm on board, and happy to update this PR to include that if @rmustacc and @rzezeski are also on board. I don't see how we do (2) without avoid conflicts, since the bits just aren't in our favor.

Reply:
Yup (1) is what I had in mind.

Reply:
Excellent. I'll expect to update this PR with another chunk that includes a VNI in the database record for a VPC. One question I actually have is about the scope. A VNI being unique to Project is probably my first intuition, but since this will appear on all underlay traffic in the rack, regardless of Project, that may not be enough. What is the scope of the VNI?

Reply:
I believe the scope of a VNI is AZ-wide.

Reply:
It seems like there is some good fodder for VNI management discussion here that perhaps deserves its own huddle. But I do want to be clear that a VNI is equivalent to a VPC -- they are 1:1.

Reply:
There are a few things here I'd like to address: regionality, relationship between UUIDs and VNIs, and assignments of IDs Also, I think this is fodder for the new RFD I mentioned above with MAC address management (and ties into things that @smklein was starting to broach in RFD 63 recently).

### Regionality

So, today in the initial bits of [RFD 24 Multi-Rack Oxide Deployments](https://rfd.shared.oxide.computer/rfd/0024), if you look at §4.4.3 and §4.4.5 the basic initial idea (which to be clear is quite rough) is that the VPC is something that intended to exist in the context of a region. However, a single VPC Subnet will be limited to an single availability-zone (in my original theory). This is meant to help limit the amount of information we need, as in theory we could have cross-AZ gateways that help limit the amount of state (ala a hoverboard). I dunno here, this is all a lot out there.

### Relationship between VPCs and Geneve IDs

So there's a bunch of back and forth here on the relationship between a VPC UUID and the Geneve ID. We need to go down the path of (1). Critically this information needs to go everywhere, so it can't just be something that we can determine based on some property. In particular, as @rcgoodfellow points out, we can't transform and truncate a 128-bit ID into a 24-bit ID. We need to allocate an ID as part of the creation of the VPC.

### But wait, 0x90, Boundary Services and more

It's very tempting to try and say, oh, boundary services is only a single thing. We only need to think about a single special Geneve ID and can just allocate everything else. But unfortunately life isn't quite that simple. There are actually a couple of different things to think about where we're going to be using Geneve IDs (and VPC features) for things that aren't quite a project today:

* We're going to have various services of ours leveraging external addresses and boundary services. This includes things like:
    * Nexus providing the API
    * External DNS
    * The console (though today that's generally out of Nexus, but could easily be on its own external address)
    * Probably some other services that I'm forgetting right now. While the initial bits of the product can't really leverage OPTE/other bits, as we add other internal services, they may eventually be able to be built in terms of the standard developer API, which would cause things.
* Today Boundary Services only uses a single ID, it'll probably want more in the future depending on how we break things up, possibly. It's not very clear to me if this is realistic or theoretical. In particular, how we do integration of VPNs and related.

So all of this tells me that we're going to need to think about how we allocate IDs here and track them. Some IDs will be used for VPCs, some will be used for internal purposes. This leads to a few thoughts:

* I think we should try to randomize IDs to try and avoid the you can tell how many VMs I've allocated.
* I don't think splitting the space here makes as much sense as we don't know how to use it.
* Given this random + scan bit from earlier, we should probably consider what this pattern looks more generally; however, I do worry about overall database performance of this and not degrading over time.
* We probably want a dedicated table that is mapping Geneve ID -> Consumer.